src/basicfuns.jl

```diff
@@ -152,9 +152,27 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
+
+Note: different than Maechler (2012), also uses bounds specific to Float32 and Float16.
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    if x > _log1pexp_thresh(x)
+        return float(x)
+    else
+        t = log1p(exp(-abs(x)))
+        return x ≤ 0 ? t : t + x
+    end
+end
+
+# threshold `thresh` such that `log1pexp(x) == x` for `x > thresh`
+_log1pexp_thresh(::Float64) = 33.27106466687738
+_log1pexp_thresh(::Float32) = 14.556091f0
+_log1pexp_thresh(::Float16) = Float16(6.24)
+_log1pexp_thresh(::BigFloat) = 172.5936479594263820448907982430859654507995334557035582760493223638550118704546
```

Comment:
I don't think this is correct in general? Shouldn't this depend on the precision of the BigFloat input?

Reply:
I'm not sure. Let me think a little more about this I think there can be a generic way to compute this threshold instead of hard-coding it.